### ISSUE TITLE
[bug] useContractRead not returning key/value pairs after second page refresh

### ISSUE THREAD

**Issue by zawilliams** [tmm/wagmi#460]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.3.5

### Current Behavior

Hey all,

I was using `useContractWrite` to write something to my local Hardhat, and then refreshed the page which triggers `useContractRead` and noticed that first refresh I get the following (note the key/value pairs):

<img width="508" alt="Screen Shot 2022-05-14 at 12 52 01 PM" src="https://user-images.githubusercontent.com/1509011/168443219-a5428b13-7339-4d07-97ab-99cb5bf0e86c.png">

And then I refresh the page again, and I get this (no key/value pairs) in the console:

<img width="513" alt="Screen Shot 2022-05-14 at 12 52 13 PM" src="https://user-images.githubusercontent.com/1509011/168443236-874de7bc-d3e6-41ce-9abd-6b618f94a6a7.png">

Is there something obvious I'm missing as to why the key/value pairs disappear after a couple refreshes when I'm returning an array of structs? Or is this a bug?

Also not sure if this is an `ethers` issue or if it's a `wagmi` issue.

### Expected Behavior

The expected behavior is for it to return the struct with key/value pairs like so:

<img width="508" alt="Screen Shot 2022-05-14 at 12 52 01 PM" src="https://user-images.githubusercontent.com/1509011/168443219-a5428b13-7339-4d07-97ab-99cb5bf0e86c.png">

**Comment by zawilliams** [comment]
Adding a link to the repo and the demo app.

Demo app: https://screech.vercel.app/

Repo with line of code in question: https://github.com/zawilliams/screech/blob/main/hooks/useScreech.js#L20

If you view the console on the page upon first load you should see an array logged a few times, but if you refresh the page, the key/value pairs inside the array disappear. You have to reload the page in an incognito window to see it happen again.

**Comment by tmm** [comment]
@zawilliams thanks for providing all this info! can you put together an example that is more minimal? will make it much easier to check this out.

**Comment by zawilliams** [comment]
@tmm had time to dig into this more. It looks like this is related to caching. If after the second refresh, you delete the `wagmi.cache` key from `localStorage`, the key/value pairs will print to the console again correctly from `useContractRead`, but then refresh again and after the data is persisted to `localStorage`, the key/value pairs get stripped.

This is where the cache is created for `wagmi.cache`:
https://github.com/tmm/wagmi/blob/44a884b84171c418f57701e80ef8de972948ef0b/packages/react/src/context.tsx#L41

Will have more time to dig in again later. Maybe it's an issue with the serializer?

**Comment by zawilliams** [comment]
Upon looking at the cached data in `wagmi.cache` in `localStorage`, it does look like the data getting persisted is indeed stripped of the key/value pairs. So, it does look like it's an issue upon writing to `localStorage` and not upon reading the cached data.

**Comment by jxom** [comment]
Thank you for looking into this. Raised a PR for a fix #552.

**Comment by zawilliams** [comment]
@jxom glad to help! Thanks for raising the PR!

**Comment by zawilliams** [comment]
@jxom @tmm I've updated my CodeSandbox (https://codesandbox.io/s/ethers-vs-wagmi-d6tq9z?file=/src/App.tsx) to use the new release (0.4.10) after the PR was merged for this issue, but it seems it has not fixed the issue. `wagmi.cache` still stores the arrays without the key/value pairs and it seems a bug has been introduced.

If you pull up the console, you'll see this now:

<img width="872" alt="Screen Shot 2022-06-08 at 1 28 52 PM" src="https://user-images.githubusercontent.com/1509011/172690609-b195bf27-edc8-481d-984c-ff41855b6c9f.png">

Note the new 3rd array key/value pair (null: Array). Previously this was just 2 arrays without the missing key/value pairs that Ethers returns correctly.

Again, here's what Ethers returns, and I assume `useContractRead` from `wagmi` _should_ return:

<img width="1523" alt="Screen Shot 2022-06-08 at 1 29 02 PM" src="https://user-images.githubusercontent.com/1509011/172691063-c75b01ed-07fb-43ca-ac7b-7ef9d0dc9f7b.png">

Should I raise a new issue or can this issue be re-opened?
